overlord/ifacestate/apparmorprompting/noticebackend.go

```diff
@@ -0,0 +1,644 @@
+		// ID maps to a notice which doesn't actually exist in userNotices.
+		// This should never occur.
+		return nil, nil, false, 0, fmt.Errorf("internal error: notice ID maps to notice which doesn't exist in user notices: %v not in %v", notice, userNotices)
+	}
+
+	return userNotices, notice, true, existingIndex, nil
+}
+
+// assembleNewUserNotices returns a new slice of notices by discarding any
+// expired notices from the front of the given userNotices, appending the given
+// notice to the end of the slice, and if it was an existing notice, shifting
+// other non-expired notices to the left to fill its original place.
+//
+// Returns the newly-assembled slice of notices, along with a closure to return
+// the original slice if there is need to roll back the changes.
+//
+// The slice is reused if possible, and memcpy is avoided unless there is need
+// to reallocate a new slice with more capacity.
+func assembleNewUserNotices(userNotices []*state.Notice, expiredCount int, notice *state.Notice, existing bool, existingIndex int) (newUserNotices []*state.Notice, restore func() []*state.Notice) {
+	// Define basic restore if the original slice isn't mutated
+	restore = func() []*state.Notice {
+		return userNotices
+	}
+	// Handle simple cases where original slice doesn't need to be mutated
+	if expiredCount == len(userNotices) {
+		// All expired (or none to begin with), so return new slice
+		newUserNotices = []*state.Notice{notice}
+		return newUserNotices, restore
+	}
+	if !existing {
+		newUserNotices = append(userNotices[expiredCount:], notice)
+		return newUserNotices, restore
+	}
+	if existingIndex == len(userNotices)-1 {
+		newUserNotices = userNotices[expiredCount:]
+		return newUserNotices, restore
+	}
+	// We now know the notice existed and was not last in the slice, and realloc
+	// will not occur since the slice length will not change.
+	newUserNotices = append(userNotices[expiredCount:existingIndex], userNotices[existingIndex+1:]...)
+	newUserNotices = append(newUserNotices, notice)
+	// Original slice was muted, so we need better restore function.
+	restore = func() []*state.Notice {
+		// Shift the originally-later notices right by one
+		copy(userNotices[existingIndex+1:], userNotices[existingIndex:])
+		userNotices[existingIndex] = notice
+		return userNotices
+	}
+	return newUserNotices, restore
+}
+
+// ntbFilter is a simplified version of state.NoticeFilter which only contains
+// information relevant to a noticeTypeBackend.
+type ntbFilter struct {
+	UserID     *uint32
+	Keys       []string
+	After      time.Time
+	BeforeOrAt time.Time
+}
+
+// simplifyFilter creates a new simplified filter with only the information
+// relevant to this backend. If no notices can match this backend, returns false.
+func (ntb *noticeTypeBackend) simplifyFilter(filter *state.NoticeFilter) (simplified *ntbFilter, matchPossible bool) {
+	if filter == nil {
+		return &ntbFilter{}, true
```

Comment:
```suggestion
		return nil, true
```

to utilize the optimization in `ntbFilter.filterNotices`

```golang
    if f == nil {
		return notices
	}
```